### PR TITLE
Adding a query parameter to filter out active alerts

### PR DESCRIPTION
**Pull request by rajagopalanand**
What does this PR do?

In some cases, we only care about the rules themselves and not the active alerts. For example, when we want to check if the rules are loaded, we do not want all the alerts but want only the rules. This PR introduces another filter to exclude active alerts from the results

Since this PR's commit history is messed up, I created another PR https://github.com/prometheus/prometheus/pull/12999 with a clean commit history. I can cancel this one

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -1414,6 +1425,12 @@ func (api *API) rules(r *http.Request) apiFuncResult {
 				if !returnAlerts {
 					break
 				}
+				var activeAlerts []*Alert
+				if excludeAlerts {
+					activeAlerts = []*Alert{}
+				} else {
+					activeAlerts = rulesAlertsToAPIAlerts(rule.ActiveAlerts())
+				}
```
Comment (yeya24) [inline web/api/v1/api.go:1433]: Can simplify this to

```
var activeAlerts []*Alert
if !excludeAlerts {
  activeAlerts = rulesAlertsToAPIAlerts(rule.ActiveAlerts())
}
```

File: web/api/v1/api.go [hunk 2]
```diff
@@ -1373,6 +1373,17 @@ func (api *API) rules(r *http.Request) apiFuncResult {
...
 
+	excludeAlertsParam := strings.ToLower(r.URL.Query().Get("exclude_active_alerts"))
+
+	if excludeAlertsParam == "" {
+		excludeAlertsParam = "false"
+	}
+
+	excludeAlerts, err := strconv.ParseBool(excludeAlertsParam)
+	if err != nil {
+		excludeAlerts = false
```
Comment (yeya24) [inline web/api/v1/api.go:1384]: Would it be better if we return 400 incase of parsing failure?

File: web/api/v1/api.go [hunk 3]
```diff
@@ -1373,6 +1373,17 @@ func (api *API) rules(r *http.Request) apiFuncResult {
 	returnAlerts := typ == "" || typ == "alert"
 	returnRecording := typ == "" || typ == "record"
 
+	excludeAlertsParam := strings.ToLower(r.URL.Query().Get("exclude_active_alerts"))
+
+	if excludeAlertsParam == "" {
+		excludeAlertsParam = "false"
+	}
+
```
Comment (qinxx108) [inline web/api/v1/api.go:1381]: nit: should we put those parsing logic into a function?

File: web/api/v1/api.go (hunk 2 above)
Comment (qinxx108) [reply web/api/v1/api.go:1384]: +1

File: web/api/v1/api.go (hunk 1 above)
Comment (rajagopalanand) [reply web/api/v1/api.go:1425]: Done

File: web/api/v1/api.go (hunk 2 above)
Comment (rajagopalanand) [reply web/api/v1/api.go]: Done

File: web/api/v1/api.go (hunk 1 above)
Comment (roidelapluie) [reply web/api/v1/api.go:1425]: no we should return nil

File: web/api/v1/api.go [hunk 4]
```diff
@@ -1462,6 +1471,20 @@ func (api *API) rules(r *http.Request) apiFuncResult {
 	return apiFuncResult{res, nil, nil, nil}
 }
 
+func parseExcludeAlerts(r *http.Request) (bool, error) {
+	excludeAlertsParam := strings.ToLower(r.URL.Query().Get("exclude_active_alerts"))
+
+	if excludeAlertsParam == "" {
+		excludeAlertsParam = "false"
```
Comment (roidelapluie) [inline web/api/v1/api.go:1478]: return false, nil

File: web/api/v1/api.go [hunk 5]
```diff
@@ -1462,6 +1471,20 @@ func (api *API) rules(r *http.Request) apiFuncResult {
...
+	}
+
+	excludeAlerts, err := strconv.ParseBool(excludeAlertsParam)
+	if err != nil {
+		return false, errors.Errorf("not supported value %q", excludeAlertsParam)
```
Comment (roidelapluie) [inline web/api/v1/api.go:1483]:
```suggestion
		return false, errors.Errorf("error converting exclude_active_alerts: %w", err)
```

File: web/api/v1/api_test.go [hunk 6]
```diff
@@ -442,12 +474,14 @@ func TestEndpoints(t *testing.T) {
 		})
 		require.NoError(t, err)
 
-		var algr rulesRetrieverMock
```
Comment (roidelapluie) [inline web/api/v1/api_test.go:445]: why?

**Review by roidelapluie — COMMENTED** [review]
why do we need so much changes in the tests?

**Comment by rajagopalanand** [comment]
> why do we need so much changes in the tests?

In order to test if active alerts are returned, active alerts slice should not be empty. Since there are no public mutators, the only other way I could think of was to actually run the rules. Once I run the rules, assertions start to fail on timestamps and other attributes that change during rule execution. Therefore I introduced a zeroFunction that would set those to zero values before assertion. All of these changes added up. If there are other ideas, I'm open to it and would prefer reducing the amount of changes

File: CHANGELOG.md [hunk 7]
```diff
@@ -3,6 +3,7 @@
 ## 2.47.1 / 2023-10-04
 
 * [BUGFIX] Fix duplicate sample detection at chunk size limit #12874
+* [ENHANCEMENT] Add ability to filter out active alerts in `/rules` API
```
Comment (roidelapluie) [inline CHANGELOG.md:6]: please remove the changelog entry

**Comment by roidelapluie** [comment]
I would just call the flag exclude_alerts since it is the name of the field, otherwise LGTM

**Comment by beorn7** [comment]
As I understand, #12999 supersedes this on. Closing.

**Pull request closed without merging**